[Zheng ZiKang] iP

zzkzzzz opened the pull request:
# Meep
> 道阻且长，行则将至

Meep ***frees your mind*** of having to remember things you need to do. It's,

- text-based
- easy to learn
-~~easy~~ ***super easy*** to use

 All you need to do is,

1. download it from [here](#).
2. double-click it.
3. add your tasks(deadline/todo/event).
4. you can manage your task through `list`/`mark`/`unmark`/`delete`.
5. exit the app by entering `bye`.


Features:

- [x] Managing tasks
   - [x] list all tasks/by specified date
   - [x] mark/unmark task
   - [x] delete task


A simple data file(`data.txt`) will be generated :grinning: :
```
T | 1 | read book
D | 1 | return book | 02/12/2019 18:00
E | 1 | project meeting | 03/12/2019 18:00
T | 1 | join sports club
D | 0 | assignment | 02/12/2019 18:00
```

zzkzzzz committed:
Implement Greet, Echo, Exit

zzkzzzz committed:
Implement Add and List tasks

zzkzzzz committed:
Use a class to represent tasks

zzkzzzz committed:
Implement Fast input reader class

zzkzzzz committed:
Implement mark and unmark task methods

zzkzzzz committed:
Add support for tracking ToDos, Events, Deadlines

As there are multiple types of tasks that have some similarity between
them. Todo, Deadline and Event classes can inherit from Task class.

Todo, Deadline and Event classes have some common behaviors
(e.g. mark() unmark() toString()).

Extracting the common behavior into a super class Task allows us to use
polymorphism when dealing with Task attributes.

zzkzzzz committed:
Add automated testing

An automated test case can be run programmatically and the result of the
test case (pass or fail) is determined programmatically. Compared to
manual testing, automated testing reduces the effort required to run
tests repeatedly and increases precision of testing

Store the test input in the text file input.txt

Store the output you expect in another text file EXPECTED.TXT

Store the script to run the test in runtest.sh/runtest.bat

Refer to this tutorial for Test using the I/O redirection technique
https://se-education.org/guides/tutorials/textUiTesting.html

zzkzzzz committed:
Implement MeepException and handle possible errors

Exceptions provide the means to separate the details of what to do when
something out of the ordinary happens from the main logic of a program.

MeepException handles invalid command, invalid date/time format and
invalid task number.

zzkzzzz committed:
Implement delete task

zzkzzzz committed:
Use Enumerations to represent the user command

zzkzzzz committed:
Implement save and restore task from text file

zzkzzzz committed:
Implement dates/time parsing and list by date

Store deadline/event dates as a java.time.LocalDate in the task objects.
Accept dates in a format dd/MM/yyyy HH:mm format (e.g., 2/12/2019 18:00)
and print in a different format MMM d yyyy hh:mm a e.g., (Oct 15 2019 06:00 PM).

Also add new command for list. User can list task before specific date.
For example "list 09/12/2019 18:00", It will list deadline/event before this date.

zzkzzzz committed:
Merge branch 'branch-Level-8' into master

zzkzzzz committed:
Extract Ui,Command,Storage,Parser,TaskList classes

Refactor the code to extract out closely related code as classes.
1. Extract Ui class to deal with interactions with the user
2. Extract Commnad class to  deal with user command
3. Extract Storage class to deal with data file restore and save
4. Extract Parser class to  deal with parsing user input
5. Extract TaskList class to  deal with actions on tasklist like add/delete

zzkzzzz committed:
Divide classes into packages

Use packages to avoid name conflicts, and to write a better maintainable code.

1.Group all Command classes under package commands.
2.Group all task classes under package task.
3.Put invalidInputException class under package exception.
4.Put Storage class under package storage.
5.Put Parder class under package parser.
6.Put Ui and Messages classes under package ui.

zzkzzzz committed:
Refactor Parser Class code

Extracting the parseUserInput() into different functions. For example,
checkTaskFormat() to check the task description format.

zzkzzzz committed:
Add JUnit tests

Add JUnit tests to test the behavior of the code.
1. test all commands class execute function
2. test Parser class with different input
3. test Storage class with vliad and invalid file path
4. test ListTask class with add/delete/generate mehtods

zzkzzzz committed:
Add JavaDoc comments

Placing Human Readable Descriptions inside of program detailing what
the Code is doing. Proper use of commenting can make code maintenance
much easier, as well as helping make finding bugs faster. Further,
commenting is very important when writing functions that other people
will use.

zzkzzzz committed:
Tweak the code to comply with Java coding standard

The aim of a coding standard is to make the entire code
base look like it was written by one person.

Coding standards help in the development of software programs
that are less complex and thereby reduce the errors.

zzkzzzz committed:
Implement find task function

Give users a way to find a task by searching for a keyword.

Add another command class - FindCommand.
Add findTasks() method to ListTask Class.
Change Parser Class to parse find command.

zzkzzzz committed:
Refactor main class structure

Separates the main function into different parts

zzkzzzz committed:
Set up gradle for build, run and unit tests

zzkzzzz committed:
Use checkStyle to detect coding style violations

putaojuice commented on src/main/java/meep/commands/UnMarkCommand.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Marks a task as undone in task list.
+ */
+public class UnMarkCommand extends Command {
```
Shouldn't it be named `UnmarkCommand`?

putaojuice commented on src/main/java/meep/task/Task.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class Task {
+    private final String title;
+    private Boolean done;
```
Shouldn't this boolean variable to be named as `isDone` ?

putaojuice reviewed (commented):
LGTM 👍 Overall I think your code follows the coding guide closely! Just some naming issues maybe you missed out.

zzkzzzz committed:
Refactor code to follow Jaca code standards

zzkzzzz committed:
Add a GUI to Meep

Use the JavaFX technology to implement the GUI.

hanhuiice commented on src/main/java/meep/TextUi/Ui.java:
```diff
@@ -0,0 +1,92 @@
+     *
+     * @param message the message.
+     */
+    public void printMsg(String... message) {
```
Clean syntax, great use of the varargs features.

hanhuiice commented on src/main/java/meep/TextUi/Ui.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Print bye.
+     */
+    public void printBye() {
```
Method names are meaningful, can get an idea of what they will do.

hanhuiice commented on src/main/java/meep/commands/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+ * Finds tasks match given keyword in task list.
+ */
+public class FindCommand extends Command {
+    public static final String COMMAND_WORD = "find";
```
Constants are all in upper case and their names provide details of their role.

hanhuiice commented on src/main/java/meep/parser/Parser.java:
```diff
@@ -0,0 +1,213 @@
+
+        String[] arr = userInput.split(" ", 2);
+        int commandLen = arr.length;
+        switch (arr[0]) {
```
The syntax for the switch statement follows the given guideline. 👍

hanhuiice reviewed (commented):
The code is clean and can follow the logic with ease.

zzkzzzz committed:
Improve the UI and move textUi to seperate package

Features added:
1. User dialog has different color with bot dialog
2. Make command input box and send button into circle
3. Change the color of the Windows
4. Change the image of user and bot

zzkzzzz committed:
Use assert to document important assumptions

An assertion failure indicates a possible bug in the code because
the code has resulted in a program state that violates an assumption
about how the code should behave. An assertion failure indicates a
possible bug and the sooner the execution stops, the safer it is.

Assertions are used in following places:
* check the string return from checkEmptyTask() to make sure it's not empty
* check the index return from parseListIndex() to make sure it's not out of range
* check format return from parseTaskFormat() to mark sure the format is correct

zzkzzzz committed:
Merge pull request #1 from zzkzzzz/branch-A-Assertions

Use assert to document important assumptions

zzkzzzz committed:
Improve code quality

Critically examines the code and refactor to improve the code quality where necessary.

Following techniques are implemented:
1. Avoid deep nesting
2. Avoid complicated expressions
3. Avoid magic numbers
4. KISS("keep it simple, stupid")
5. SLAP hard
6. Make the happy path prominent
7. Delete dead cdode

zzkzzzz committed:
Improve code quality

Critically examines the code and refactor to improve the code
quality where necessary.

Following techniques are implemented:
1. Avoid deep nesting
2. Avoid complicated expressions
3. Avoid magic numbers
4. KISS("keep it simple, stupid")
5. SLAP hard
6. Make the happy path prominent
7. Delete dead code

zzkzzzz committed:
Merge pull request #2 from zzkzzzz/branch-A-CodeQuality

Branch a code quality

zzkzzzz committed:
Set up Continuous Integration using Github action

zzkzzzz committed:
Rename CI workflow folder name

zzkzzzz committed:
Fix expected test result in Junit test

zzkzzzz committed:
Add new command 'help' to guide user

This feature is under category C-Help which provide
in-App guidance to users.

User can simply type 'help' then the bot will display the
guidence. Moreover, the guidence will display when the bot start.

zzkzzzz committed:
Extract error messages to Messages class

zzkzzzz committed:
Fix import library order bug

zzkzzzz committed:
Change main class to UI Launcher

zzkzzzz committed:
Fix Junit Test error message

zzkzzzz committed:
Fix package naming issue

zzkzzzz committed:
Fix StorageTest bug in CI

zzkzzzz committed:
Remove unused imports

zzkzzzz committed:
Fix Github CI on Windows date formatting

zzkzzzz committed:
Fix find command bug

zzkzzzz committed:
Add a User Guide to the project

zzkzzzz committed:
Fix user guide typo

zzkzzzz committed:
Refactor code to improve code quality
